src/str.js

```diff
@@ -628,17 +633,121 @@ Sk.builtin.str.prototype["rindex"] = new Sk.builtin.func(function (self, tgt, st
+    }
+
+    if (start === undefined || Sk.builtin.checkNone(start)) {
+        start = 0;
+    } else {
+        start = Sk.builtin.asnum$(start);
+        start = start >= 0 ? start : self.v.length + start;
+    }
+
+    if (end === undefined || Sk.builtin.checkNone(end)) {
+        end = self.v.length;
+    } else {
+        end = Sk.builtin.asnum$(end);
+        end = end >= 0 ? end : self.v.length + end;
+    }
+
+    if(start > self.v.length){
+        return Sk.builtin.bool.false$;
+    }
+
+    var substr = self.v.slice(start, end);
+
+    if(Sk.abstr.typeName(prefix) == "tuple"){
+        var tmpBool = false, resultBool = false;
+        var it, i;
+        for (it = Sk.abstr.iter(prefix), i = it.tp$iternext(); i !== undefined; i = it.tp$iternext()) {
+            if(start > end){
+                tmpBool = start <= 0;
+            }else{     
+                tmpBool = substr.indexOf(i.v) === 0;    
+            }
+            resultBool = resultBool || tmpBool;
+            if(resultBool){
+                break;
+            }
+        }
+        return resultBool?Sk.builtin.bool.true$ : Sk.builtin.bool.false$;
+    }
+
+    if(prefix.v == "" && start > end && end >= 0){
+        return Sk.builtin.bool.false$;
+    }
+
+    return new Sk.builtin.bool(substr.indexOf(prefix.v) === 0);
 });
 
 // http://stackoverflow.com/questions/280634/endswith-in-javascript
-Sk.builtin.str.prototype["endswith"] = new Sk.builtin.func(function (self, tgt) {
-    Sk.builtin.pyCheckArgsLen("endswith", arguments.length, 2, 2);
-    Sk.builtin.pyCheckType("tgt", "string", Sk.builtin.checkString(tgt));
-    return new Sk.builtin.bool( self.v.indexOf(tgt.v, self.v.length - tgt.v.length) !== -1);
+Sk.builtin.str.prototype["endswith"] = new Sk.builtin.func(function (self, suffix, start, end) {
+    Sk.builtin.pyCheckArgsLen("endswith", arguments.length - 1, 1, 3);
+
+    if(Sk.abstr.typeName(suffix) != "str" && Sk.abstr.typeName(suffix) != "tuple"){
+        throw new Sk.builtin.TypeError("endswith first arg must be str or a tuple of str, not " + Sk.abstr.typeName(suffix));
+    }
+
+    if ((start !== undefined) && !Sk.builtin.checkInt(start) && !Sk.builtin.checkNone(start)) {
+        throw new Sk.builtin.TypeError("slice indices must be integers or None or have an __index__ method");
+    }
+    if ((end !== undefined) && !Sk.builtin.checkInt(end) && !Sk.builtin.checkNone(end)) {
+        throw new Sk.builtin.TypeError("slice indices must be integers or None or have an __index__ method");
+    }
+
+    if (start === undefined || Sk.builtin.checkNone(start)) {
+        start = 0;
+    } else {
+        start = Sk.builtin.asnum$(start);
+        start = start >= 0 ? start : self.v.length + start;
+    }
+
+    if (end === undefined || Sk.builtin.checkNone(end)) {
+        end = self.v.length;
+    } else {
+        end = Sk.builtin.asnum$(end);
+        end = end >= 0 ? end : self.v.length + end;
+    }
```

Comment:
For most cases this is not a problem but since the error message suggests that start and end can be instances of a class with an ‘__index__‘ method it might be more correct to use misceval.isIndex and misceval.asIndex rather than checkInt and asnum